File: 0303-Range Sum Query - Immutable/0303-Range Sum Query - Immutable.cpp
class NumArray {
public:
    NumArray(vector<int> nums) {
        int n = nums.size();
        sums = vector<int>(1 + n);
        for (int i = 1; i <= n; ++i) {
            sums[i] = sums[i - 1] + nums[i - 1];
        }
    }
    
    int sumRange(int i, int j) {
        return sums[j + 1] - sums[i];
    }
    
private:
    vector<int> sums;
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(i,j);
 */
 

File: 0301-0400/0303-Range Sum Query - Immutable/0303-Range Sum Query - Immutable.py
class NumArray:
    
    def __init__(self, nums: List[int]):
        self.presum = [0] * (len(nums) + 1)
        for i, num in enumerate(nums):
            self.presum[i + 1] = self.presum[i] + num

    def sumRange(self, i: int, j: int) -> int:
        return self.presum[j + 1] - self.presum[i]


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(i,j)
